src/StoreApi/Routes/Checkout.php

```diff
@@ -317,6 +330,75 @@ protected function create_or_update_draft_order() {
 		return $order_object;
 	}
 
+	/**
+	 * Check request options and store (shop) config to determine if a user account
+	 * should be created as part of order processing.
+	 *
+	 * @param \WP_REST_Request $request The current request object being handled.
+	 *
+	 * @return boolean True if a new user account should be created.
+	 */
+	protected function should_create_customer_account( $request ) {
+		if ( is_user_logged_in() ) {
+			return false;
+		}
+
+		$checkout_requires_account = true === filter_var( get_option( 'woocommerce_checkout_registration_required' ), FILTER_VALIDATE_BOOLEAN );
+		if ( $checkout_requires_account ) {
+			return true;
+		}
+
+		$user_requested_account = ! empty( $request['create_account'] ) && true === filter_var( $request['create_account'], FILTER_VALIDATE_BOOLEAN );
+		if ( $user_requested_account ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Create a new account for a customer.
+	 *
+	 * @throws RouteException If an error is encountered when creating the user account.
+	 *
+	 * @param string $user_email The email address to use for the new account.
+	 * @param string $first_name The first name to use for the new account.
+	 * @param string $last_name  The last name to use for the new account.
+	 *
+	 * @return int User id if successful
+	 */
+	protected function create_customer_account( $user_email, $first_name, $last_name ) {
+		$customer_id = wc_create_new_customer(
+			$user_email,
+			'',
+			'',
+			array(
+				'first_name' => $first_name,
+				'last_name'  => $last_name,
+			)
+		);
+
+		if ( is_wp_error( $customer_id ) ) {
+			switch ( $customer_id->get_error_code() ) {
+				case 'registration-error-email-exists':
+					throw new RouteException(
+						'woocommerce_rest_checkout_create_account_failure',
+						__( 'An account is already registered with your email address. Please log in.', 'woo-gutenberg-products-block' ),
```

Comment:
I'd love to show a login link when this error happens. I don't think we currently support this in the [notices system](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/docs/block-client-apis/notices.md), so if we were to implement that it would need to be follow up. 

Keen for others thoughts on this, doesn't feel like an urgent priority, but could be a nice enhancement to the notices system and make the checkout flow smoother for that case :)

Reply:
Looks like the woo core version of this error message has a link handled by javascript: https://github.com/woocommerce/woocommerce/blob/e736ac4145d9aceb3574922d01824ede546a07f3/includes/wc-user-functions.php#L47

Reply:
Related: #2854